Попов Захар

BlizPerfect opened the pull request:
@masssha1308

Примеры работы программы и результаты выполнения прошлых частей задания по ссылкам:
1. 1 часть: https://github.com/kontur-courses/tdd/pull/257
2. 2 часть: https://github.com/kontur-courses/di-updated/pull/18

BlizPerfect committed:
Добавлены проекты из TagCloudll

BlizPerfect committed:
Старые классы, выбрасывающие исключения, переведены на паттерн Result<T>

BlizPerfect committed:
Рефакторинг кода FontParser.

FontParser теперь возвращает FontFamily

BlizPerfect committed:
Бизнес логика из контейнера выделена в отдельные классы.

1. Добавлена фабрика по созданию WordFilter.
2. Добавлена фабрика по созданиюCloudLayouterWorker.
3. Рефакторинг DIContainer.

BlizPerfect committed:
Добавлены новые аргументы для командной строки.

1. wordsToIncludeFile - добавление слов в фильтр "скучных слов" из файла.
2. wordsToExcludeFile- исключение слов из фильтра "скучных слов" из файла.
3. normalizeMinCoefficient - минимальный коэффициент нормализации.
3. Рефакторинг WordFilter.

BlizPerfect committed:
Общий рефакторинг кода по результатам проверки второй части задания TagCloud

BlizPerfect committed:
Рефакторинг CloudLayouterPainter.

1. Рефакторинг CloudLayouterPainter.
2. Добавлен новый тест.

BlizPerfect committed:
Рефакторинг CircularCloudLayouter.

BlizPerfect committed:
Рефакторинг NormalizedFrequencyBasedCloudLayouterWorker.

BlizPerfect committed:
Рефакторинг RandomCloudLayouterWorker.

BlizPerfect committed:
Рефакторинг ImageSaver.

1. Рефакторинг ImageSaver.
2. Обновление тестов.

BlizPerfect committed:
Рефакторинг Normalizer.

1. Рефакторинг Normalizer.
2. Обновление тестов.

BlizPerfect committed:
Добавлены тесты для BoolParser.

BlizPerfect committed:
Рефакторинг ColorParser.

1. Рефакторинг ColorParser.
2. Добавлены тесты для ColorParser.

BlizPerfect committed:
Рефакторинг FontParser.

1. Рефакторинг FontParser.
2. Добавлены тесты для FontParser.

BlizPerfect committed:
Рефакторинг SizeParser.

1. Рефакторинг SizeParser.
2. Добавлены тесты для SizeParser.

BlizPerfect committed:
Рефакторинг TagCloud.

BlizPerfect committed:
Добавлены тесты для проверка корректности переданных аргументов.

BlizPerfect committed:
Рефакторинг тестов.

masssha1308 commented on FileSenderRailway/Result.cs:
```diff
@@ -1,5 +1,8 @@
 ﻿using System;
+using System.Runtime.CompilerServices;
 
+[assembly: InternalsVisibleTo("TagCloud")]
```
А для чего? Заметила что это нужно для обращения к Value класса Result. Но ведь это поле не просто так было сделано internal :) Например, вот здесь есть проверка что при вычислении ширины и высоты не было ошибок
![image](https://github.com/user-attachments/assets/ac65743c-c61e-4b51-9f38-9839ed5964d5)
Так почему бы вместо прямого доступа к Value не воспользоваться GetValueOrThrow?

BlizPerfect replied:
Это **только** для того, что бы класс Result.cs был виден для моих проектов. Можно было перекопировать этот класс в сам проект, но я решил просто сделать его видимым.

В работе программы нигде GetValueOrThrow не используется, потому что, как я понял, в этом и смысл задания - исключения не должны выбрасываются явно. Если же мы используем GetValueOrThrow, то у нас выбросится исключение на неверных данных.

masssha1308 replied:
> Это только для того, что бы класс Result.cs был виден для моих проектов. Можно было перекопировать этот класс в сам проект, но я решил просто сделать его видимым.

Не совсем понятно, почему бы публичный класс из другого проекта не будет виден без этого атрибута?)

masssha1308 commented on TagCloud.Tests/CloudLayouterPaintersTest/CloudLayouterPainterTest.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace TagCloud.Tests.CloudLayouterPaintersTest
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Interoperability", "CA1416:Проверка совместимости платформы", Justification = "<Ожидание>")]
```
`System.Diagnostics.CodeAnalysis.SuppressMessage` длинновато, давай сократим с помощью юзинга

masssha1308 replied:
А вообще, для чего этот атрибут здесь и что значит "ожидание"?

BlizPerfect replied:
Это подавление сообщения об ограниченности использования класса Bitmap:
![image](https://github.com/user-attachments/assets/1dcd4a76-3e39-470c-9490-60f88a82a6df)

Можно было сделать так:
```
#pragma warning disable CA1416
            var dummyImage = new Bitmap(1, 1);
#pragma warning restore CA1416
```

Изначально добавлял игнорирование этой ошибки в GlobalSuppressions.cs:
```
// This file is used by Code Analysis to maintain SuppressMessage
// attributes that are applied to this project.
// Project-level suppressions either have no target or are given
// a specific target and scoped to a namespace, type, member, etc.

using System.Diagnostics.CodeAnalysis;

[assembly: SuppressMessage("Interoperability", "CA1416:Проверка совместимости платформы", Justification = "<Ожидание>", Scope = "member", Target = "~M:TagCloud.Tests.ImageSaversTests.ImageSaverTests.SaveFile_ThrowsException_WithInvalidFormat(System.String)")]
```

Justification обозначает вроде как причину, почему это подавление добавлено. Студия по умолчания туда прописывает "<Ожидание>" - ожидается. что когда-то там исправят. Я понимаю, что это плохая практика и в продакшене так конечно не надо делать.

masssha1308 replied:
> ожидается. что когда-то там исправят

Ожидается) Так может и поправим?)

masssha1308 commented on TagCloud.Tests/CloudLayouterPaintersTest/CloudLayouterPainterTest.cs:
```diff
@@ -0,0 +1,52 @@
+namespace TagCloud.Tests.CloudLayouterPaintersTest
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Interoperability", "CA1416:Проверка совместимости платформы", Justification = "<Ожидание>")]
+    internal class CloudLayouterPainterTest
```
Кажется тут все тесты с одинаковым смыслом - невалидный параметр на входе и проверяется что выкидывается исключение с нужным текстом. Давай объедим все это дело в один тест с testcases. Тогда и setUp можно будет убрать. И еще не хватает проверки позитивных сценариев. Не только что все "как надо" ломается, а еще и что работает правильно

masssha1308 commented on TagCloud.Tests/CloudLayouterTests/CircularCloudLayouterTests/CircularCloudLayouterMainRequirementsTest.cs:
```diff
@@ -0,0 +1,188 @@
+{
+    [TestFixture]
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Interoperability", "CA1416:Проверка совместимости платформы", Justification = "<Ожидание>")]
+    internal class CircularCloudLayouterMainRequirementsTest
```
Давай объединим CircularCloudLayouterMainRequirementsTest и CircularCloudLayouterTest. Не вижу смысла в выделении отдельного класса для теста с exception. + смущает "MainRequirements" в названии, я бы тоже убрала (если не согласен, то опиши зачем оно)

BlizPerfect replied:
Я согласен, что конкретно сейчас название "CircularCloudLayouterMainRequirementsTest" не котируется, но оно появилось из первой части задания, как выделение проверки основным требованиям:
![image](https://github.com/user-attachments/assets/db7245f5-6cb7-40fb-9725-44d78f41e76a)

Из работы с прошлым наставником, я лично подчерпнул, что надо не бояться создавать множество тестовых файлов. Это поможет хорошо структурировать их. И конкретно тут, мне нравится, как идейно сделано, ещё со времён первой части задания - у нас есть класс, отвечающий за основные требования к раскладке и всё остальное, куда вошла только проверка на исключение.

masssha1308 replied:
> появилось из первой части задания, как выделение проверки основным требованиям

Люди которые в дальнейшем работали бы с кодом этого приложения не знали бы об этих требованиях и этом ТЗ. Поэтому для них логика в названии бы отсутствовала. Нужно поправить

masssha1308 replied:
> Из работы с прошлым наставником, я лично подчерпнул, что надо не бояться создавать множество тестовых файлов

Нужно исправить т.к. "основные" - понятие непонятное :) хранители знания что такое в данном контексте "основное" - только люди которые видели ТЗ к задаче. А у человека который видит этот код впервые это вызовет недоумение. Следовательно, это мешает читабельности кода

masssha1308 replied:
А так, да. Создавать тестовые файлы не нужно бояться **если есть на то причина**. Очень советую почитать 9 главу "Микросевисы" Ричардсона

masssha1308 commented on TagCloud.Tests/CloudLayouterTests/CircularCloudLayouterTests/CircularCloudLayouterMainRequirementsTest.cs:
```diff
@@ -0,0 +1,188 @@
+
+namespace TagCloud.Tests.CloudLayouterTests.CircularCloudLayouterTests
+{
+    [TestFixture]
```
Где-то есть TestFixture, а где-то нет. Хочется придерживаться единого стиля. А вообще, в используемой версии nunit TestFixture [необязателен](https://docs.nunit.org/articles/nunit/writing-tests/attributes/testfixture.html#:~:text=Beginning%20with%20NUnit%202.5%2C%20the%20TestFixture%20attribute%20is%20optional%20for%20non%2Dparameterized%2C%20non%2Dgeneric%20fixtures.%20So%20long%20as%20the%20class%20contains%20at%20least%20one%20method%20marked%20with%20the%20Test%2C%20TestCase%20or%20TestCaseSource%20attribute%2C%20it%20will%20be%20treated%20as%20a%20test%20fixture.)

masssha1308 commented on TagCloud.Tests/CloudLayouterTests/CircularCloudLayouterTests/CircularCloudLayouterMainRequirementsTest.cs:
```diff
@@ -0,0 +1,188 @@
+            var maxRectangleWidth = 70;
+            var minRectangleHeight = 20;
+            var maxRectangleHeight = 50;
+            var rectanglesCount = 1000;
```
почему именно такие значения? может быть их тоже лучше рандомайзером генерить?

BlizPerfect replied:
На мой взгляд, рандомить границы, для того, что бы затем рандомить размер внутри этих границ - оверкил в данном случае, так как в этом тесте мы проверяем соответствие основным требованиям из первой части задания для ICloudLayouter - форма, плотность, отсутствие пересечений прямоугольников. А эти значения используются для вспомогательного класса, которые и выдаёт размеры прямоугольников, которые затем будут расставлены.

masssha1308 commented on TagCloud.Tests/CloudLayouterTests/CircularCloudLayouterTests/CircularCloudLayouterMainRequirementsTest.cs:
```diff
@@ -0,0 +1,188 @@
+            rectangles = tags.Select(x => x.Rectangle).ToArray();
+        }
+
+        [TestCase(0.7, 1000)]
```
Есть ли смысл выносить testCase когда он один?

masssha1308 commented on TagCloud.Tests/CloudLayouterTests/CircularCloudLayouterTests/CircularCloudLayouterMainRequirementsTest.cs:
```diff
@@ -0,0 +1,188 @@
+            }
+        }
+
+        private (int start, int end) GetGridIndexesInterval(
```
Хорошо что вынес в отдельные методы, так намного читабельнее

masssha1308 commented on TagCloud.Tests/ImageSaversTests/ImageSaverTest.cs:
```diff
@@ -0,0 +1,83 @@
+            Directory.CreateDirectory(directoryPath);
+        }
+
+        [SetUp]
```
а точно ли нужен такой setUp? В чем преимущество перед тем чтобы вместо imageSaver в тестах писать new ImageSaver()?

masssha1308 commented on TagCloud.Tests/MainTest.cs:
```diff
@@ -0,0 +1,60 @@
+namespace TagCloud.Tests
+{
+    [TestFixture]
+    internal class MainTest() : BaseOptionTest("MainTest")
```
Неинформативное название

masssha1308 commented on TagCloud.Tests/NormalizersTest/NormalizerTest.cs:
```diff
@@ -0,0 +1,83 @@
+namespace TagCloud.Tests.NormalizersTest
+{
+    [TestFixture]
+    internal class NormalizerTest
```
В некоторых классах в названии не tests, а test. Давай все такие места поправим на tests

BlizPerfect replied:
А как можно закоммитить изменения названия файла, без того, что бы гит удалил старый файл и добавил новый, с новым названием и содержимым старого?

Я думал этот коммит не вызовет этого:
![image](https://github.com/user-attachments/assets/23799c4a-434c-48dc-bc54-0d5aa2284c90)

> git mv old_filename new_filename

Но в более автоматическом режиме, в графическом интерфейса Студии(Или райдера например) как такое сделать?

masssha1308 commented on TagCloud.Tests/OptionsTests/BackgroundColorTests.cs:
```diff
@@ -0,0 +1,27 @@
+namespace TagCloud.Tests.OptionsTests
+{
+    [TestFixture]
+    internal class BackgroundColorTests() : BaseOptionTest("BackgroundColor")
```
А зачем выделять для этого теста отдельный класс, почему его нельзя поместить в MainTest? Кажется по смыслу он туда подходит. Классы стоит выделать по функциональности которую проверяем (в данном случае это тест на ProgramExecutor) поэтому на мой взгляд нет необходимости проверку фона при работе выносить в отдельный класс. Если класс получится слишком большим и будет напрашиваться разделение, можно выделить 2: негативные и позитивные сценарии

masssha1308 replied:
Тот же самый вопрос к классам TextColorTest, FontTest и т.д.

BlizPerfect replied:
Как я писал выше, на мой взгляд это очень повышает читаемость кода и результатов теста. Даже в обозревателе тестов выглядит гармонично:
![image](https://github.com/user-attachments/assets/a0f53471-0fb5-4dc5-aeff-ce6df4d9b3bf)

- Название папки - OptionsTests - тестируются аргументы командной строки, передаваемые нашей программе.
- Название файла/Класса - FontTests - тестируются опции, относящиеся к шрифтам, которые мы передаём как аргументы для работы нашей программы.
- Название теста - Program_WorksCorrectly_WithInvalidFont - тестируется работа программы, при передаче через командную строку некорректного аргумента, относящегося к шрифтам.

Из работы с прошлым наставником я такую логику структурирования подчерпнул.

Название "MainTest" - откровенно не лучшее, я честно говоря думал, что поменял его. Сейчас его название лучше отражает, что я имел ввиду - FullProgramExecutionTests - тесты, направленные на проверку результата полной работы программы.

masssha1308 commented on TagCloud.Tests/OptionsTests/ValidValues.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    internal static class ValidValues
+    {
+        public readonly static string[] ValidDataFileContent = new string[]
```
inconsistent modifiers declaration order

masssha1308 commented on TagCloud.Tests/WordCountersTests/WordCounterTest.cs:
```diff
@@ -0,0 +1,46 @@
+                { "Three", 1 },
+                { "Four", 4 },
+            };
+            var values = new string[]
```
Кажется такое есть в ValidValues

masssha1308 commented on fp.sln:
```diff
@@ -1,14 +1,21 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
```
кажется в mr попали лишние файлы

BlizPerfect committed:
Рефакторинг сообщения об ограниченности поддержки Bitmap на разных платформах.

BlizPerfect committed:
Рефакторинг теста для CloudLayouterPainter

BlizPerfect committed:
Во всех тестовых классах добавлен недостающий атрибут [TestFixture]

BlizPerfect committed:
Атрибут [TestCase] у тестов с одним случаем исправлен на просто [Test]

BlizPerfect committed:
Рефакторинг тестов с избыточным атрибутом [SetUp]

BlizPerfect committed:
MainTest переименован в FullProgramExecution, для большей информативности

BlizPerfect committed:
Исправлен порядок модификаторов

BlizPerfect committed:
Исправелно дублирование кода в WordCounterTest

BlizPerfect committed:
Названия тестов приведены к единообразию

BlizPerfect committed:
Добавлено 2 новых теста

BlizPerfect committed:
Замена использования internal свойства Value у Result<T> на публичный метод GetValueOrThrow()

BlizPerfect committed:
Добавлено сообщение о ограниченности выполнения кода

BlizPerfect committed:
Рефакторинг CircularCloudLayouterTest.

1. Тесты с CircularCloudLayouterMainRequirementsTest добавлены в CircularCloudLayouterTest.
2. Рефакторинг CircularCloudLayouterTest.
